agent/src/util/discovery_operator.rs

```diff
@@ -0,0 +1,1411 @@
+/// methods are being used. Therefore, #[allow(dead_code)] has been added to all methods that are not invoked or
+/// tested on a DiscoveryOperator.
+#[derive(Clone)]
+pub struct DiscoveryOperator {
+    discovery_handler_map: RegisteredDiscoveryHandlerMap,
+    config: KubeAkriConfig,
+    instance_map: InstanceMap,
+    device_plugin_path: String,
+}
+
+#[cfg_attr(test, automock)]
+impl DiscoveryOperator {
+    pub fn new(
+        discovery_handler_map: RegisteredDiscoveryHandlerMap,
+        config: KubeAkriConfig,
+        instance_map: InstanceMap,
+        device_plugin_path: String,
+    ) -> Self {
+        DiscoveryOperator {
+            discovery_handler_map,
+            config,
+            instance_map,
+            device_plugin_path,
+        }
+    }
+    /// Returns discovery_handler_map field. Allows the struct to be mocked.
+    #[allow(dead_code)]
+    pub fn get_discovery_handler_map(&self) -> RegisteredDiscoveryHandlerMap {
+        self.discovery_handler_map.clone()
+    }
+    /// Returns config field. Allows the struct to be mocked.
+    #[allow(dead_code)]
+    pub fn get_config(&self) -> KubeAkriConfig {
+        self.config.clone()
+    }
+    /// Returns instance_map field. Allows the struct to be mocked.
+    #[allow(dead_code)]
+    pub fn get_instance_map(&self) -> InstanceMap {
+        self.instance_map.clone()
+    }
+    /// Returns device_plugin_path field. Allows the struct to be mocked.
+    #[allow(dead_code)]
+    pub fn get_device_plugin_path(&self) -> String {
+        self.device_plugin_path.clone()
+    }
+    #[allow(dead_code)]
+    pub async fn stop_all_discovery(&self) {
+        let mut discovery_handler_map = self.discovery_handler_map.lock().unwrap().clone();
+        if let Some(protocol_dhs_map) =
+            discovery_handler_map.get_mut(&self.config.spec.protocol.name)
+        {
+            for (endpoint, dh_details) in protocol_dhs_map.clone() {
+                match dh_details.stop_discovery.send(()) {
+                    Ok(_) => trace!("stop_all_discovery - discovery client for protocol {} at endpoint {} told to stop", self.config.spec.protocol.name, endpoint),
+                    Err(e) => error!("stop_all_discovery - discovery client for protocol {} at endpoint {} could not receive stop message with error {:?}", self.config.spec.protocol.name, endpoint, e)
+                }
+            }
+        }
+    }
+
+    /// Calls discover on the Discovery Handler at the given endpoint and returns the connection stream.
+    pub async fn get_stream(
+        &self,
+        endpoint: &str,
+        discovery_handler_path: &str,
+    ) -> Option<StreamType> {
+        let discover_request = tonic::Request::new(DiscoverRequest {
+            discovery_details: self.config.spec.protocol.discovery_details.clone(),
+        });
+        trace!("get_stream - endpoint is {}", endpoint);
+        if endpoint == EMBEDDED_DISCOVERY_HANDLER_ENDPOINT
```

Comment:
this code makes sense to me (parse and compare endpoint to known embedded, path  of internal uds, else try external) ... but this might be easier to understand if we used something like an enum to distinguish between embedded, uds, and external.

Reply:
Are you thinking of something like a `get_stream_location` function that is called at the beginning of `get_stream` and returns an enum, which would be matched for in the function?
```
pub enum StreamLocation {
    Embedded,
    Uds,
    Network,
}

pub fn get_stream_location(
    endpoint: &str,
    discovery_handler_path: &str,
) -> Result<StreamLocation, Err<url::ParseError>> {
    if endpoint == EMBEDDED_DISCOVERY_HANDLER_ENDPOINT {
        StreamLocation::Embedded
    } else if endpoint.starts_with(discovery_handler_path) {
        StreamLocation::Uds
    } else {
        match Url::parse(endpoint) {
            Ok(_) => StreamLocation::Network,
            Err(e) => Err(e),
        }
    }
}
```

Reply:
essentially yes ... my thought was that it would be better not to have to parse endpoint at all ... that wherever `endpoint` is set (i think maybe in the env vars of the DH pod?), another variable (like `endpoint_type`) could be set that identified how to interpret `endpoint`.

Reply:
`endpoint` is passed as a part of a RegistrationRequest as a `str`. Discovery handlers pass the socket or ip address they are running on in this `RegisterRequest`. The registration service (that receives the requests) could set the enum; then, the key of the `ProtocolDiscoveryHandlerMap` (in `register.rs`) would be an enum instead of a str representation of the endpoint, which does seem cleaner.

Reply:
fixed in 861b05a. Now, when `register` receives a request it checks if the endpoint String maps to one of the `DiscoveryHandlerEndpoint` enums. If not, it returns an error. Before this, a bad endpoint wouldnt be noticed until a Configuration was applied. Definitely needed to be fixed. Thanks!